### PR TITLE
[VAL] Add register testing infrastructure and ifc_reg tests

### PR DESCRIPTION
**Pull request by wkkuna**
* Allow for logging code reduction with custom define
* Add 'ifc_reg' library for register testing infrastructure
* Add tests:
  * ifc_reg_rand_test
  * ifc_reg_ro_reg_rand_test
  * ifc_reg_soc_rw_cptra_ro_config
  * ifc_reg_soc_rw_cptra_ro_idevid
  * ifc_reg_soc_rw_cptra_ro_pk_hash
  * ifc_reg_soc_rw_cptra_ro_misc
  * ifc_reg_intr_test
  * ifc_reg_dma_soc_access
  * ifc_reg_rw_hw
  * ifc_reg_soc_rw_hw_ro
  * ifc_reg_invalid_access

Depends on https://github.com/chipsalliance/caliptra-rtl/pull/1321

### CODE REVIEW

File: src/integration/test_suites/libs/printf/printf.h [hunk 1]
```diff
@@ -46,6 +46,20 @@ int printf(const char* format, ...);
         printf(fmt, ##__VA_ARGS__); \
     }
 
+#define VPRINTF_FATAL(format, ...) VPRINTF(FATAL, format, ##__VA_ARGS__)
```
Comment (Nitsirks) [inline src/integration/test_suites/libs/printf/printf.h]: The intent of VERBOSITY is to be able to run regressions with minimal logging. This isn't necessary, you can just run regressions with VERBOSITY set to LOW and accomplish the same thing.

Comment (wkkuna) [reply src/integration/test_suites/libs/printf/printf.h]: The issue here was not limiting the print verbosity.

After having added the `soc_access` library over the main branch we noticed that the sizes of some of the test binaries exceeded the permitted ROM size. (Not limited to the newly added tests but also the pre-existing ones, specifically the extensive DMA ones)

I.e. the linker would fail with:
```
ERROR: ROM size exceeds 96KiB
```

The compiler we'd use is the same one referenced in the workflow files: https://github.com/chipsalliance/caliptra-rtl/blob/main/.github/workflows/build-test-verilator.yml#L90

Setting `verbosity_g` at the test level doesn't get rid of the actual code that's contained within other libraries. Hence bloating the final binary.

New macros were introduced to ensure that no code is produced for the logging above a certain threshold. Now I see it could be improved if instead of an arbitrary threshold the `CPT_VERBOSITY` was used to determine the logging macro implementation.

The first step we performed was to, of course, slim down the `soc_access`  library itself but this alone was not sufficient to guarantee the size requirement for all tests.

We hesitated to modify any optimization options for the compiler as this could lead to a less fine-control of the test execution (which is crucial in testing some of the edge-case scenarios).

We also considered re-working the way the tests are compiled (to exclude the libraries that are not used for each test scenario) - but due to the way they're intertwined we decided that introducing a new approach for logging is the most sensible way forward.

Of course, if you have a dedicated process to deal with the test binary size I will gladly revert the logging changes and adapt the preferred way.

Comment (Nitsirks) [reply src/integration/test_suites/libs/printf/printf.h]: We have a .mk file solution already to add new libs to individual tests that need them instead of every test. See the iccm tests as an exemple: chipsalliance/caliptra-rtl/src/integration/test_suites/directed_test_iccm_hash_overflow/directed_test_iccm_hash_overflow.mk

I *think* this will help solve your issue in a better way than modifying every single test.

**Comment by wkkuna** [comment]
Hi @calebofearth,

Thank you for the merge of `combined-pr-antmicro`.

I see that some of the PRs that I'm actively addressing have been closed:
* https://github.com/chipsalliance/caliptra-rtl/pull/1297
* https://github.com/chipsalliance/caliptra-rtl/pull/1311
* https://github.com/chipsalliance/caliptra-rtl/pull/1321
* https://github.com/chipsalliance/caliptra-rtl/pull/1322
* https://github.com/chipsalliance/caliptra-rtl/pull/1323
* https://github.com/chipsalliance/caliptra-rtl/pull/1324

Would you like me to re-open them rebased and retested?

Thanks!

**Comment by Nitsirks** [comment]
We wanted to get the PRs that were done with reviews merged, I've reopened the remaining PRs and opened a fresh combined branch for you to continue targeting. Please address the remaining comments on these PRs, and resolve merge conflicts. We will work on getting them merged asap.

Also, please update your PR names to follow the conventions detailed here: https://github.com/chipsalliance/caliptra-rtl/blob/main/CONTRIBUTING.md#pull-request-conventions

File: src/integration/test_suites/libs/printf/printf.h (hunk 1 above)
Comment (wkkuna) [reply src/integration/test_suites/libs/printf/printf.h]: You were correct, it was not reasonable to ask for such a significant structural change, especially within a singular PR coupled with other changes.

I ended up proceeding as instructed (with the .mk files for additional libraries). It didn't resolve the binary size issue completely.
Especially those tests that included both the `soc_access` and `ifc_reg` binaries exceeded the ROM size after this change alone.

I went over the libraries again, reduced logging and some QoL features and they finally fit within the requirement.
The functionality stayed the same - it just requires a bit more legwork in debugging when something fails.
